every_election/apps/elections/baker.py

```
@@ -1,2 +1,89 @@
-def push_event_to_queue():
-    print("trigger_rebuild")
+import json
+import logging
+import os
+from typing import Dict, Optional
+
+import boto3
+from django.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+def event_bus_exists(event_bus_arn: str) -> bool:
```

This is all to guard against us typo-ing the evenbus we're trying to send events to. We check it exists before we send an event. This is because events will fail silently if an event bus doesn't exist.

I don't really like this solution, because it feels quite brittle, having the specific `DC_CHECK_EVENTBUS_ROLE` in the account with the eventbus we're targetting, but I haven't come up with anything better. The alternative would be just to delete this check.

---

..and then, assuming the issue is that pushing an event fails silently (I assume that must be the issue, otherwise we'd just catch the error) is there any other operation we can try that would fail. For example, could we try a `describe_event_bus` before we `put_events`? Or does that also require a assuming a different Iam role? We're assigning `DescribeEventBus` in https://github.com/DemocracyClub/EveryElection/blob/a2c4b9b7a4427d1fdde66e8ae412f190a5552b57/cdk_stacks/stacks/code_deploy_policies.py#L109-L119

---

I've just realised the last commit you pushed was switching from `list_event_buses` to `describe_event_bus` :facepalm:

---

Yeah no worries.

Just had a quick test with `put_events` and the value of `Entries` doesn't contain anything helpful if the ARN was bad, which is quite annoying. So its basically we do this check, or we don't.

I think I'd be fine with saying it is on us to set the ARN correctly just to simplify the code here. Might be a good one to get a second opinion from @symroe on.

I think if we _are_ going to do this check, we should probably move it though. I reckon we should check the bus exists one on app startup in a system check, rather than every time we push an event to it (because of performance impact).

Lets decide if we actually want to do it first though.

---

I don't think the response does help us. 
This is the response to an event bus that doesn't exist: 
```
{'FailedEntryCount': 0,
 'Entries': [{'EventId': '0dcb7c51-2c36-87ee-39e2-fa56a0855f41'}],
 'ResponseMetadata': {'RequestId': '9a1f38a7-16ed-443e-a81f-e093e1118b21',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'x-amzn-requestid': '9a1f38a7-16ed-443e-a81f-e093e1118b21',
   'content-type': 'application/x-amz-json-1.1',
   'content-length': '85',
   'date': 'Thu, 24 Apr 2025 15:12:16 GMT'},
  'RetryAttempts': 0}}
```

Versus to one that does

```
{'FailedEntryCount': 0,
 'Entries': [{'EventId': '7ce793b7-7038-c618-52ef-b220308646ba'}],
 'ResponseMetadata': {'RequestId': '5397ef8c-492e-4233-af01-e59f6b52579d',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'x-amzn-requestid': '5397ef8c-492e-4233-af01-e59f6b52579d',
   'content-type': 'application/x-amz-json-1.1',
   'content-length': '85',
   'date': 'Thu, 24 Apr 2025 15:14:04 GMT'},
  'RetryAttempts': 0}}
```

The event that worked I can see in the console. The first just gets swallowed as best I can tell.

---

Sorry didn't see you're comment :) I was doing the same. 

Good point re. doing the check once.

---

I wrote the above comment before reading all of this, and now I'm making everything more confusing :laughing: Sorry!

So, broadly we agree this is a bit of a smell, and I think we might agree that it's better implemented as a one-time check.

Another approach, that we might want to consider as part of a wider "how to even event bus" conversation, is that we could host a 'gateway' lambda in the event bus hosting account. Lambda _does_ have good cross-org permissions and will fail if it doesn't exist. Calling Lambda in async mode is non-blocking, and the Lambda will deal with retrying enqueuing failures, too.  It's also a place for us to add additional logging / whatever, if that's useful. 

An added bonus to the approach is that we'd have a single gateway that all apps could send events to, and at this single point we can add some validation of the event payload.

I'm not really suggesting we do this now (I'd like to get this merged before the elections to see how it works under load), but it might save pushing this sort of logic to the clients. 

```python
lambda_client.invoke(
    FunctionName=function_arn,
    InvocationType="Event",
    Payload={},
)
```

Is a really easy thing to call, and is basically non-blocking (we use this approach for the DC Logging system, as Firehose doesn't have good cross-org permissions either)